bundles/org.eclipse.jface/src/org/eclipse/jface/resource/DeviceResourceDescriptor.java

```diff
@@ -80,7 +82,7 @@ final boolean shouldBeCached() {
 	 * @return the newly allocated resource (not null)
 	 * @throws DeviceResourceException if unable to allocate the resource
 	 */
-	public abstract Object createResource(Device device);
+	public abstract R createResource(Device device);
```

Comment:
Only changed the `createResource()` method to return a generic type R and not `destroyResource()` to accept a generic type T, because this would mean that in the sub-classes `destroyResource()` would have to be changed to only accept `Image/Color/Font`.
But narrowing a method parameter type is AFAIK not a binary compatible change (but would probably be source-compatible in most cases), while it should be binary compatible to narrow the return type.

Reply:
After double checking the assumption that narrowing the return type of a method is binary-compatible I learned, that this assumption is wrong... (for example [this article](https://www.morling.dev/blog/refining-return-type-java-methods-without-breaking-backwards-compatibility/) explains it and and I tried it out myself).

So I think we have two options now:

1. Assume that nobody besides `ResourceManager` actually calls `DeviceResourceDescriptor.createResource()` respectively the overrides because all three sub-classes for Images, Colors and Fonts have a corresponding `createImage/Color/Font()` method and that specific method is preferred because it returns the specify resource type. In my SDK workspace I see no other caller.
If we desides this is ok, I have to check what happens to custom sub-classes of DeviceResourceDescriptor.
The question is does anybody knows other code that calls createResource() or uses custom `DeviceResourceDescriptor`?
2.  Create a new method e.g. `createTypedResource()` that uses the generic return type and deprecate the old method.
3. Don't use the generic type parameter for this method and just stick with Object. Then the new generic parameter of the `DeviceResourceDescriptor` class would only be used by the `ResourceManager` and not DeviceResourceDescriptor itself.